Práctica 1.1 — Probabilidad y Estadística 
Análisis de la evolución poblacional mundial

**Actividad 1 — Identificación de la fuente**

Los datos utilizados en esta práctica provienen de Our World in Data (OWID).
La fuente original es United Nations, World Population Prospects (2024).

**Cita:**
> UN, World Population Prospects (2024) – processed by Our World in Data.
> "Annual change in population – UN WPP" [dataset]. United Nations,
> "World Population Prospects"; United Nations, "World Population Prospects
> - Interim Update" [original data].

**Preguntas:**

1. **¿Quién produjo originalmente los datos?**
   La División de Población de la ONU (UN DESA), mediante el informe
   *World Population Prospects (2024)*.

2. **¿Qué organización los procesa para su uso en OWID?**
   Our World in Data (Global Change Data Lab, Universidad de Oxford)
   estandariza y limpia los datos crudos de la ONU.

3. **¿Qué representa la variable que analizarás?**
   La columna "Annual change in population" representa el cambio neto
   de población, calculado como la diferencia entre la población al 1 de julio de años consecutivos.
   Está expresada en número de personas y cubre el periodo histórico 1951–2023.


**Actividad 2 — Carga y exploración de datos**

In [1]:
#2.1 Importar las bibliotecas
import pandas as pd
import numpy as np

In [197]:
#2.2 Cargar el archivo CSV
df = pd.read_csv("../data/raw/annual-population-growth/annual-population-growth.csv")

In [198]:
#2.3 Explorar el DataFrame(df)
# --> primeras filas
df.head()

,Entity,Code,Year,Annual change in population,Annual population change (Projected)
0,Afghanistan,AFG,1951,103163.0,NaN
1,Afghanistan,AFG,1952,108441.0,NaN
2,Afghanistan,AFG,1953,108919.0,NaN
3,Afghanistan,AFG,1954,111251.0,NaN
4,Afghanistan,AFG,1955,119027.0,NaN


In [199]:
# --> dimensiones (cuantas filas y cuantas columnas) 
df.shape

(38400, 5)

In [200]:
# --> nombres de columnas
df.columns.tolist()

['Entity',
 'Code',
 'Year',
 'Annual change in population',
 'Annual population change (Projected)']

In [201]:
# --> tipos de datos
df.dtypes

Entity                                      str
Code                                        str
Year                                      int64
Annual change in population             float64
Annual population change (Projected)    float64
dtype: object

In [202]:
# --> estadísticas descriptivas 
# std = desviacion estantar
df.describe()

,Year,Annual change in population,Annual population change (Projected)
count,38400.000000,1.868800e+04,1.971200e+04
mean,2025.500000,2.063019e+06,8.375481e+05
std,43.300872,9.440859e+06,5.816745e+06
min,1951.000000,-3.315935e+06,-2.291035e+07
25%,1988.000000,1.781000e+03,-9.742750e+03
50%,2025.500000,4.689800e+04,9.350000e+01
75%,2063.000000,3.021128e+05,1.411160e+05
max,2100.000000,9.337137e+07,7.206736e+07


In [203]:
# --> cantidad de valores faltantes
df.isnull().sum()

Entity                                      0
Code                                     1200
Year                                        0
Annual change in population             19712
Annual population change (Projected)    18688
dtype: int64

In [204]:
# --> cantidad de entidades
df["Entity"].nunique() 

256

**Preguntas:**

1. **¿Cuántas filas y columnas tiene el dataset?**
   38,400 filas y 5 columnas.

2. **¿Qué tipo de datos contiene la variable Year?**
   Entero (int64).

3. **¿Existen valores faltantes?**
   Sí: 1,200 en Code, 19,712 en Annual change in population y 18,688 en Annual population change (Projected). 
   Los faltantes en las dos últimas se explican porque una columna es para años históricos y la otra para proyectados.

4. **¿Cuántas entidades diferentes contiene el dataset?**
   256 entidades (incluye países individuales y agregados regionales/mundiales).

**Actividad 3 — Estadística descriptiva**

In [205]:
# Hecho en clase
df.head(3)

,Entity,Code,Year,Annual change in population,Annual population change (Projected)
0,Afghanistan,AFG,1951,103163.0,NaN
1,Afghanistan,AFG,1952,108441.0,NaN
2,Afghanistan,AFG,1953,108919.0,NaN


In [206]:
# Hecho en clase
df_year = df[["Year"]] #--> (doble [[]] son dos dimensiones una fila y una columnas = df) (un [] es una serie)
df_year.head()

,Year
0,1951
1,1952
2,1953
3,1954
4,1955


In [207]:
# Hecho en clase
#  Filtrar por filas --> La entidad (columan) YEAR en un tiempo especifico
df_extraccion_filas = df[df["Year"] == 2020]
df_extraccion_filas.head()

,Entity,Code,Year,Annual change in population,Annual population change (Projected)
69,Afghanistan,AFG,2020,1212852.0,NaN
219,Africa (UN),UN_AFR,2020,32815675.0,NaN
369,Albania,ALB,2020,-13061.0,NaN
519,Algeria,DZA,2020,747543.0,NaN
669,American Samoa,ASM,2020,-451.0,NaN


In [208]:
# Hecho en clase
#  Filtrar dos entidades con un valor especifico 
df_2020_mex = df[(df["Code"] == "MEX") & (df["Year"] == 2020)]
df_2020_mex.head()

,Entity,Code,Year,Annual change in population,Annual population change (Projected)
22269,Mexico,MEX,2020,1036075.0,NaN


In [209]:
# Hecho en clase
#  Filtrar dos entidades con más de un valor especifico 
df_2005_2020_mex = df[(df["Code"] == "MEX") & (df["Year"] >= 2005) & (df["Year"]<= 2020)]
df_2005_2020_mex.head() #--> Los primeros 5
# df_2005_2020_mex #--> Todos

,Entity,Code,Year,Annual change in population,Annual population change (Projected)
22254,Mexico,MEX,2005,1417373.0,NaN
22255,Mexico,MEX,2006,1442162.0,NaN
22256,Mexico,MEX,2007,1520690.0,NaN
22257,Mexico,MEX,2008,1599930.0,NaN
22258,Mexico,MEX,2009,1625434.0,NaN


In [210]:
# Hecho en clase
# Extraer solo 3 entidades especificas
df_extraccion_rango_mex_EYA = df_2005_2020_mex [["Entity","Year","Annual change in population"]]
df_extraccion_rango_mex_EYA.head()

,Entity,Year,Annual change in population
22254,Mexico,2005,1417373.0
22255,Mexico,2006,1442162.0
22256,Mexico,2007,1520690.0
22257,Mexico,2008,1599930.0
22258,Mexico,2009,1625434.0


In [211]:
# Hecho en clase
df_extraccion_rango_mex_EYA.mean(numeric_only=True)

Year                              2012.500
Annual change in population    1400307.625
dtype: float64

In [246]:
# Seleccionar un año del periodo historico
df_year = df[(df["Year"] >= 2005) & (df["Year"] <= 2023) & (df["Entity"] == "Honduras")]

In [247]:
df_hist = df_year[['Annual change in population']]

In [239]:
media = df_year.mean()
print('Media:', media)

Media: Annual change in population    2.063019e+06
dtype: float64


In [214]:
mediana = df_year.median()
print('Mediana:', mediana)

Mediana: Annual change in population    175977.0
dtype: float64


In [215]:
varianza = df_year.var()
print('Varianza:', varianza)

Varianza: Annual change in population    6.164734e+06
dtype: float64


In [216]:
desviacion = df_year.std()
print('Desviación estándar:', desviacion)

Desviación estándar: Annual change in population    2482.888175
dtype: float64


In [217]:
minimo = df_year.min()
print('Mínimo:', minimo)

Mínimo: Annual change in population    170243.0
dtype: float64


In [218]:
maximo = df_hist.max()
print('Máximo:', maximo)

Máximo: Annual change in population    180973.0
dtype: float64


In [253]:
# Entidades correspondientes al minimo y maximo de todo el df
entidad_min = df.loc[df["Annual change in population"].idxmin(),"Entity"]
entidad_max = df.loc[df["Annual change in population"].idxmax(),"Entity"]
print("Valor mínimo:", entidad_min)
print("Valor máximo:", entidad_max)

Valor mínimo: Ukraine
Valor máximo: World


**Preguntas:**

1. **¿Cuál es el valor de la media?**
   2.063019e+06

2. **¿Cuál es el valor de la mediana?**
   175977.0

3. **¿Cuál es el valor de la desviación estándar?**
   2482.888175

4. **¿Qué entidad presenta el valor mínimo y cuál el máximo?**
   Valor mínimo es Ukraine y el máximo es World

**Actividad 4 — Comprobación del entorno**

In [220]:
# Extraer la version de los paquetes
import sys

print("Python:", sys.version)
print("Pandas:", pd.__version__)
print("NumPy:", np.__version__)

Python: 3.14.7 (tags/v3.14.7:823f032, Aug  5 2026, 10:51:32) [MSC v.1944 64 bit (AMD64)]
Pandas: 3.0.5
NumPy: 2.5.2


#Hecho en clase
Indexadores iloc y loc 

Loc (significa location utiliza etuiquetas)--> seleccionar por etiquetas/nombres
- df.loc[filas, columnas]
- Ej: df.iloc[10] --> Regresa la fila que ocupa la posición 10
- Ej: df.iloc[:,3] --> Regresa todas las filas de la columna que ocupa la posión 3
- Se ocupan valores numericos  pero tambien texto
- Busca el valos en índice en las filas

Iloc  (significa integer location)--> seleccionar por posiciones numericas 
- df.iloc[filas, columnas]
- Ej: df.iloc[10] --> Regresa la fila cuya etiqueta de índice es igual a 10
- Ej: df.iloc[:,"Year"] --> Regresa todas las filas de la colmna cuyo nombre es Year
- Se ocupan valores numericos nada más
- Busaca la posicíon en las filas

In [221]:
df.loc[:,["Entity", "Year"]] #:, es para las filas

,Entity,Year
0,Afghanistan,1951
1,Afghanistan,1952
2,Afghanistan,1953
3,Afghanistan,1954
4,Afghanistan,1955
...,...,...
38395,Zimbabwe,2096
38396,Zimbabwe,2097
38397,Zimbabwe,2098
38398,Zimbabwe,2099


In [222]:
df.iloc[30:35,0:12] 

,Entity,Code,Year,Annual change in population,Annual population change (Projected)
30,Afghanistan,AFG,1981,-1231727.0,NaN
31,Afghanistan,AFG,1982,-946206.0,NaN
32,Afghanistan,AFG,1983,-73395.0,NaN
33,Afghanistan,AFG,1984,272237.0,NaN
34,Afghanistan,AFG,1985,236633.0,NaN


In [223]:
# Ejercicio en clase
# 1.- loc - filas: Selcciona la fila cuyo  
df.loc[10000]
#df.iloc[10000]
#Da lo mismo por que concide con el valor del indice y las posicicones tambien en 0

Entity                                  Equatorial Guinea
Code                                                  GNQ
Year                                                 2051
Annual change in population                           NaN
Annual population change (Projected)              46407.0
Name: 10000, dtype: object

In [224]:
# Ejercicio en clase
# 2.- loc - Columnas. Selecciona todas las filas y únicamente las columnas Entity, Year y Annual change in population
df.loc[:,["Entity","Year","Annual change in population"]] #--> :, quiero todas las filas y columnas

,Entity,Year,Annual change in population
0,Afghanistan,1951,103163.0
1,Afghanistan,1952,108441.0
2,Afghanistan,1953,108919.0
3,Afghanistan,1954,111251.0
4,Afghanistan,1955,119027.0
...,...,...,...
38395,Zimbabwe,2096,NaN
38396,Zimbabwe,2097,NaN
38397,Zimbabwe,2098,NaN
38398,Zimbabwe,2099,NaN


In [225]:
# Ejercicio en clase
# 3.- Combina ambos ejercicios (1.-loc y 2.-iloc)  
df.loc[10000,["Entity","Year","Annual change in population"]] #--> los : no porque ya tiene un indice deinido

Entity                         Equatorial Guinea
Year                                        2051
Annual change in population                  NaN
Name: 10000, dtype: object

In [226]:
# Ejercicio en clase
# 4.- loc - Filtro: Seleciona todos los registros de Mexico correspondientes al año 2020  
df.loc[(df["Entity"] == 'Mexico') & (df["Year"] == 2020)]

,Entity,Code,Year,Annual change in population,Annual population change (Projected)
22269,Mexico,MEX,2020,1036075.0,NaN


In [257]:
# Ejercicio en clase
# 5.- iloc - Filas: Selecionar las primeras 5 filas de DataFrame
df.iloc[0:5]

,Entity,Code,Year,Annual change in population,Annual population change (Projected)
0,Afghanistan,AFG,1951,103163.0,NaN
1,Afghanistan,AFG,1952,108441.0,NaN
2,Afghanistan,AFG,1953,108919.0,NaN
3,Afghanistan,AFG,1954,111251.0,NaN
4,Afghanistan,AFG,1955,119027.0,NaN


In [259]:
# Ejercicio en clase
# 6.- iloc - Columnas: Seleciona todas las filas de las columnas que ocupan las posiciones 0,2 y 3
df.iloc[["0","2","3"]]

,Entity,Code,Year,Annual change in population,Annual population change (Projected)
0,Afghanistan,AFG,1951,103163.0,NaN
2,Afghanistan,AFG,1953,108919.0,NaN
3,Afghanistan,AFG,1954,111251.0,NaN


In [258]:
# Ejercicio en clase
# 7.- iloc - Filas y columnas: Seleccionar las filas de la posición 0 a la 4 y las columnas de la posición 0 a la 3
df.iloc[0:5,0:4] 

,Entity,Code,Year,Annual change in population
0,Afghanistan,AFG,1951,103163.0
1,Afghanistan,AFG,1952,108441.0
2,Afghanistan,AFG,1953,108919.0
3,Afghanistan,AFG,1954,111251.0
4,Afghanistan,AFG,1955,119027.0
